metrics.yaml

```diff
@@ -0,0 +1,43 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# This file defines the metrics that are recorded by the Glean SDK.
+# APIs to use these pings are automatically generated at build time using
+# the `glean_parser` PyPI package.
+
+# Metrics in this file may make use of SDK reserved ping names. See
+# https://mozilla.github.io/glean/book/dev/core/internal/reserved-ping-names.html
+# for additional information.
+
+---
+$schema: moz://mozilla.org/schemas/glean/metrics/2-0-0
+
+user:
+  attention:
+    type: event
```

Comment:
I read [the current docs](https://github.com/mozilla-rally/rally-study-01/blob/master/DATA_SCHEMA.md) and, by looking at the code, it seems to me that we're sending "one ping per event". Is this intentional?

The current model I implemented is an hybrid thing _only to showcase the possibilities_: it creates an event and slaps it into a ping (call `measurements`), then it sends the ping right away. I believe that a much better collection model would follow this pattern:

- have an event for each single _real event_ (e.g. page opened, attention start, attention end, ..)
- record an event to Glean when the _real event_ happens.
- submit the `measurements` ping containing all the events when needed.

This has a few nice upsides:

- it allows a more detailed analysis, since it enables reconstructing the event sequence/timeline (which is not possible in the current model especially if there's restart across the collection). 
- it gives a better understanding of what's being collected and how.
- it improves the semantics around the defined events (the word "event" means different things in WebScience, RS01)

Reply:
(I'm approaching this as a study partner not as a data person at mozilla)

"submit the `measurements` ping containing all the events _when needed"_ sounds like a lot to ask the study partner. Already I have no resources to make the right choice to this end. Are you telling me I have to figure out the scheduling on my own? Isn't Glean supposed to do this? I'm just pushing data as its ready & assuming that Rally / telemetry handles the scheduling for me.

Reply:
> "submit the `measurements` ping containing all the events _when needed"_ sounds like a lot to ask the study partner. 

It's the same thing we're asking partners to do already :-) 

> Already I have no resources to make the right choice to this end. Are you telling me I have to figure out the scheduling on my own?

Not quite (see below), but please note that you are already making scheduling decisions on your own (i.e. calling `rally.sendPing`)

> Isn't Glean supposed to do this? I'm just pushing data as its ready & assuming that Rally / telemetry handles the scheduling for me.

Well, you're currently not doing this: partners are required to (1) assemble a payload and (2) decide where to send it (the scheduling part). The current Glean.js version does not offer any built-in ping out of the box, but there's plans to provide an _events_ ping in the short term (but it's not in the currently available version). I'm not sure any other Glean-owned ping (with the exception of the `deletion-request`) would fit perfectly in the Rally-world. However, let me expand a bit on my previous comment.

We have the following possibilities.

**Exact parity to the current legacy implementation**
We define the `measurement` ping and add each field of the legacy ping [as listed here](https://github.com/mozilla-rally/rally-study-01/blob/master/DATA_SCHEMA.md) (for example pageId) as a field of this Glean ping. In addition to that, we can add two [ping reasons](https://mozilla.github.io/glean/book/user/pings/custom.html#defining-a-custom-ping) that represent audio and attention. The semantics in this case will be exactly the same as the current legacy implementation.

**Events + extras in a custom ping** (similar to what I was suggesting in the previous comment)
Instead of listening exclusively to the final event, listen to the intermediate events and record these events in Glean. Events will be modeled as Glean events and added to the custom ping. Times are computed server side by analysing the events timeline. The ping is submitted with the same schedule as RS01 currently does.

**Events + Glean 'events' ping** (not currently possible as Glean.js does not support the 'events' ping yet)
Like the previous option, except that instead of using a custom ping events are sent, automatically, in the 'events' ping.

Reply:
I think that all approaches outlined are equally valid from an analysis perspective.  However there are different challenges in the analytics flow that will require compromise.

By bundling the start and stop times of each activity into one 'event' each row will reflect a period of behavior (Hamilton pref) as opposed to a moment in time (Alessio pref).  Bundling moments in into periods by design can have analytics benefits when we already know the question we are going to ask.  Putting ourselves in the shoes of Academic partners, this might be their first encounter with click-stream analytics and it does take some education to understand how to connect the dots in a meaningful way. They don't need to consider this nuance if we take care of it through data collection design.

By collecting each discrete event we are baking in more flexibility in our analysis and are set up better for analytics at scale.  We could more easily answer several questions because the data is formatted as discrete moments where any combination could potentially be interesting, but it requires a more sophisticated understanding of how to collect these dots.  If we design telemetry in a way where we are pushing the pain of connecting the dots to the analyst, we should also provide resources on how and why.

Thus, my recommendation depends on how our product evolves.  If we are going to ask partners that are not familiar with click-stream analytics to use the data we should go with Hamilton's preference (processed moments that reflect concrete ideas like a period of time).  If we want to collect one data set to rule them all where technical partners or internal contributors connect the dots we should go with Alessio's preference.

Reply:
Thanks for chiming in @magorlick !

Very clear analysis, I like it :-) Regardless of what you folks end up picking here, make sure to keep the pros and cons in mind, with the understanding that Glean allows you to change/adapt quickly if you need to change.

Reply:
> This said, it seems that the goal is to make sure that the data we have in the glean version matches the data we have in the current implementation, so I would say we should stick as close to what these custom pings currently look like, even if there are analytical downsides to it. Given that schema changes won't be as big an issue in the future, I'm assuming that if we opt to move toward more of a clickstream-style event structure, we should be able to do that with little fuss, correct?

Perfect! I updated this PR to match the current collection. Yes, moving forward changes on collection could be deployed much faster